In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('https://query.data.world/s/3m3lsknojsj4vf6lzlichtk7ljr645?dws=00000')
df.head()

brewery_id             brewery_name  review_time  review_overall  \
0       10325          Vecchio Birraio   1234817823             1.5   
1       10325          Vecchio Birraio   1235915097             3.0   
2       10325          Vecchio Birraio   1235916604             3.0   
3       10325          Vecchio Birraio   1234725145             3.0   
4        1075  Caldera Brewing Company   1293735206             4.0   

   review_aroma  review_appearance review_profilename  \
0           2.0                2.5            stcules   
1           2.5                3.0            stcules   
2           2.5                3.0            stcules   
3           3.0                3.5            stcules   
4           4.5                4.0     johnmichaelsen   

                       beer_style  review_palate  review_taste  \
0                      Hefeweizen            1.5           1.5   
1              English Strong Ale            3.0           3.0   
2          Foreign / Export Stout            3.0           3.0   
3                 German Pilsener            2.5           3.0   
4  American Double / Imperial IPA            4.0           4.5   

                beer_name  beer_abv  beer_beerid  
0            Sausa Weizen       5.0        47986  
1                Red Moon       6.2        48213  
2  Black Horse Black Beer       6.5        48215  
3              Sausa Pils       5.0        47969  
4           Cauldron DIPA       7.7        64883

# **CLEANING THE DATA**

**Checking for duplicate rows**

In [ ]:
df.duplicated().sum()

0

**Checking for NULL values**

In [ ]:
df. isnull(). sum()


brewery_id                0
brewery_name             15
review_time               0
review_overall            0
review_aroma              0
review_appearance         0
review_profilename      348
beer_style                0
review_palate             0
review_taste              0
beer_name                 0
beer_abv              67785
beer_beerid               0
dtype: int64

In [ ]:
df1 = df[df.isna().any(axis=1)]
print (df1)

         brewery_id                 brewery_name  review_time  review_overall  \
273            1075      Caldera Brewing Company   1103668195             3.0   
430             850   Moon River Brewing Company   1110736110             3.5   
603             850   Moon River Brewing Company   1100038819             4.0   
733            1075      Caldera Brewing Company   1260673921             4.0   
798            1075      Caldera Brewing Company   1212201268             4.5   
...             ...                          ...          ...             ...   
1586568       14359  The Defiant Brewing Company   1187052567             4.0   
1586587       14359  The Defiant Brewing Company   1177842168             3.5   
1586596       14359  The Defiant Brewing Company   1287951067             4.0   
1586597       14359  The Defiant Brewing Company   1241906223             4.5   
1586598       14359  The Defiant Brewing Company   1236550020             4.0   

         review_aroma  revi

# **BEERS WITH HIGHEST ABV VALUES**

In [ ]:
df.sort_values(by=['beer_abv'], ascending=False).drop_duplicates('beer_name').head(10)

brewery_id                                       brewery_name  \
12919          6513                                       Schorschbräu   
12939          6513                                       Schorschbräu   
746385        16315                                            BrewDog   
12941          6513                                       Schorschbräu   
996153        15237                                De Struise Brouwers   
748356        16315                                            BrewDog   
12918          6513                                       Schorschbräu   
1042328         173  Hair of the Dog Brewing Company / Brewery and ...   
746421        16315                                            BrewDog   
293821           35                 Boston Beer Company (Samuel Adams)   

         review_time  review_overall  review_aroma  review_appearance  \
12919     1316780901             4.0           4.0                4.0   
12939     1309974178             4.0           4.0                3.5   
746385    1285808609             3.5           4.0                4.0   
12941     1321041551             3.0           3.0                3.0   
996153    1316875384             4.5           5.0                5.0   
748356    1299094828             4.5           4.0                3.5   
12918     1248785936             3.5           4.0                4.0   
1042328   1238562892             4.5           4.5                4.0   
746421    1318612252             4.5           4.0                2.5   
293821    1258857067             4.0           5.0                4.0   

        review_profilename                        beer_style  review_palate  \
12919          kappldav123                           Eisbock            4.0   
12939             Sunnanek                           Eisbock            4.0   
746385               bobsy    American Double / Imperial IPA            4.0   
12941            orangesol                           Eisbock            2.0   
996153           Brabander  American Double / Imperial Stout            4.5   
748356             fear025  American Double / Imperial Stout            3.5   
12918          kappldav123                           Eisbock            4.0   
1042328         SacoDeToro                English Barleywine            4.5   
746421            Sunnanek           Belgian Strong Pale Ale            4.0   
293821             WeisGuy               American Strong Ale            4.0   

         review_taste                      beer_name  beer_abv  beer_beerid  \
12919             3.5  Schorschbräu Schorschbock 57%     57.70        73368   
12939             4.0  Schorschbräu Schorschbock 43%     43.00        57856   
746385            4.0             Sink The Bismarck!     41.00        57015   
12941             3.0  Schorschbräu Schorschbock 40%     39.44        55712   
996153            5.0     Black Damnation VI - Messy     39.00        70633   
748356            4.5       Tactical Nuclear Penguin     32.00        54571   
12918             4.5  Schorschbräu Schorschbock 31%     30.86        51466   
1042328           5.0                           Dave     29.00        35197   
746421            4.0                     Ghost Deer     28.00        74021   
293821            4.5           Samuel Adams Utopias     27.00        25759   

         no_of_reviews  average_overall_review  
12919                1                4.000000  
12939                2                3.750000  
746385              76                3.197368  
12941                3                3.333333  
996153               7                4.500000  
748356              88                3.187500  
12918                1                3.500000  
1042328             16                4.437500  
746421               3                3.500000  
293821             355                4.066197

# **DETAILS OF BEER WITH HIGHEST ABV VALUE**

In [ ]:
df.loc[df['beer_abv'].idxmax()]


brewery_id                                         6513
brewery_name                               Schorschbräu
review_time                                  1316780901
review_overall                                      4.0
review_aroma                                        4.0
review_appearance                                   4.0
review_profilename                          kappldav123
beer_style                                      Eisbock
review_palate                                       4.0
review_taste                                        3.5
beer_name                 Schorschbräu Schorschbock 57%
beer_abv                                           57.7
beer_beerid                                       73368
no_of_reviews                                         1
average_overall_review                              4.0
Name: 12919, dtype: object

# **FACTORS THAT IMPACT THE QUALITY OF BEER THE MOST**

In [ ]:
df.sort_values(by=['review_overall'], ascending=False)

brewery_id              brewery_name  review_time  review_overall  \
793307         579          Kona Brewing Co.   1293809208             5.0   
591801        1549     Goose Island Beer Co.   1180667386             5.0   
591818        1549     Goose Island Beer Co.   1175185922             5.0   
591819        1549     Goose Island Beer Co.   1175135314             5.0   
591824        1549     Goose Island Beer Co.   1173843941             5.0   
...            ...                       ...          ...             ...   
653706        1412  Red Rock Brewing Company   1030008142             0.0   
37234         1413       Desert Edge Brewery   1030197670             0.0   
653695        1412  Red Rock Brewing Company   1030007859             0.0   
39667         1413       Desert Edge Brewery   1030266889             0.0   
39648         1413       Desert Edge Brewery   1030267158             0.0   

        review_aroma  review_appearance review_profilename  \
793307           2.5                3.0             vfgccp   
591801           5.0                4.5             bort11   
591818           5.0                5.0             Beezor   
591819           4.5                4.0             zeff80   
591824           4.5                3.5             klewis   
...              ...                ...                ...   
653706           3.0                0.0           beernut7   
37234            2.0                0.0           beernut7   
653695           2.0                0.0           beernut7   
39667            3.0                0.0           beernut7   
39648            4.0                0.0           beernut7   

                              beer_style  review_palate  review_taste  \
793307           American Pale Ale (APA)            4.0           3.5   
591801  American Double / Imperial Stout            5.0           4.5   
591818  American Double / Imperial Stout            4.5           5.0   
591819  American Double / Imperial Stout            4.5           5.0   
591824  American Double / Imperial Stout            5.0           5.0   
...                                  ...            ...           ...   
653706          American Amber / Red Ale            2.5           3.5   
37234                     Czech Pilsener            3.0           2.0   
653695                        Hefeweizen            2.5           2.0   
39667            American Pale Ale (APA)            2.0           2.0   
39648                    Irish Dry Stout            2.0           2.0   

                         beer_name  beer_abv  beer_beerid  
793307          Fire Rock Pale Ale       5.8         5753  
591801  Bourbon County Brand Stout      14.5        10672  
591818  Bourbon County Brand Stout      14.5        10672  
591819  Bourbon County Brand Stout      14.5        10672  
591824  Bourbon County Brand Stout      14.5        10672  
...                            ...       ...          ...  
653706          Red Rock Amber Ale       4.0         3787  
37234                     Pub Pils       4.0         3806  
653695     Red Rock Bavarian Weiss       4.0         3788  
39667                Utah Pale Ale       4.0         3810  
39648            Latter Days Stout       4.0         3804  

[1586614 rows x 13 columns]

**Finding the correlation between columns in order to identify factors that affect "review_overall"**

In [ ]:
df.corr()

<ipython-input-62-c1f0a49b9c47>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr()


brewery_id  review_time  review_overall  review_aroma  \
brewery_id                1.000000     0.220954       -0.009331      0.019069   
review_time               0.220954     1.000000        0.028724      0.067453   
review_overall           -0.009331     0.028724        1.000000      0.616013   
review_aroma              0.019069     0.067453        0.616013      1.000000   
review_appearance         0.001993     0.058233        0.501732      0.561029   
review_palate            -0.000499     0.057247        0.701914      0.616947   
review_taste             -0.002672     0.044498        0.789816      0.716776   
beer_abv                  0.063731     0.144885        0.138512      0.332570   
beer_beerid               0.424097     0.461535        0.003794      0.077669   
no_of_reviews            -0.202901    -0.059231        0.168684      0.184541   
average_overall_review   -0.014963     0.097207        0.543267      0.574703   

                        review_appearance  review_palate  review_taste  \
brewery_id                       0.001993      -0.000499     -0.002672   
review_time                      0.058233       0.057247      0.044498   
review_overall                   0.501732       0.701914      0.789816   
review_aroma                     0.561029       0.616947      0.716776   
review_appearance                1.000000       0.566634      0.546980   
review_palate                    0.566634       1.000000      0.734135   
review_taste                     0.546980       0.734135      1.000000   
beer_abv                         0.263941       0.286711      0.290827   
beer_beerid                      0.054794       0.044510      0.030402   
no_of_reviews                    0.189736       0.189383      0.207283   
average_overall_review           0.513262       0.553826      0.596044   

                        beer_abv  beer_beerid  no_of_reviews  \
brewery_id              0.063731     0.424097      -0.202901   
review_time             0.144885     0.461535      -0.059231   
review_overall          0.138512     0.003794       0.168684   
review_aroma            0.332570     0.077669       0.184541   
review_appearance       0.263941     0.054794       0.189736   
review_palate           0.286711     0.044510       0.189383   
review_taste            0.290827     0.030402       0.207283   
beer_abv                1.000000     0.201818       0.123658   
beer_beerid             0.201818     1.000000      -0.395048   
no_of_reviews           0.123658    -0.395048       1.000000   
average_overall_review  0.258130     0.008385       0.310500   

                        average_overall_review  
brewery_id                           -0.014963  
review_time                           0.097207  
review_overall                        0.543267  
review_aroma                          0.574703  
review_appearance                     0.513262  
review_palate                         0.553826  
review_taste                          0.596044  
beer_abv                              0.258130  
beer_beerid                           0.008385  
no_of_reviews                         0.310500  
average_overall_review                1.000000

**The top 3 factors are:**

-   taste- 0.789816
-   palate- 0.701914
-  aroma- 0.616013



#**TOP BEERS TO RECOMMEND TO SOMEONE**

In [ ]:
#Sorting the table to view beers with highest overall review value
df.sort_values(by=['review_overall'], ascending=False)

brewery_id              brewery_name  review_time  review_overall  \
793307         579          Kona Brewing Co.   1293809208             5.0   
591801        1549     Goose Island Beer Co.   1180667386             5.0   
591818        1549     Goose Island Beer Co.   1175185922             5.0   
591819        1549     Goose Island Beer Co.   1175135314             5.0   
591824        1549     Goose Island Beer Co.   1173843941             5.0   
...            ...                       ...          ...             ...   
653706        1412  Red Rock Brewing Company   1030008142             0.0   
37234         1413       Desert Edge Brewery   1030197670             0.0   
653695        1412  Red Rock Brewing Company   1030007859             0.0   
39667         1413       Desert Edge Brewery   1030266889             0.0   
39648         1413       Desert Edge Brewery   1030267158             0.0   

        review_aroma  review_appearance review_profilename  \
793307           2.5                3.0             vfgccp   
591801           5.0                4.5             bort11   
591818           5.0                5.0             Beezor   
591819           4.5                4.0             zeff80   
591824           4.5                3.5             klewis   
...              ...                ...                ...   
653706           3.0                0.0           beernut7   
37234            2.0                0.0           beernut7   
653695           2.0                0.0           beernut7   
39667            3.0                0.0           beernut7   
39648            4.0                0.0           beernut7   

                              beer_style  review_palate  review_taste  \
793307           American Pale Ale (APA)            4.0           3.5   
591801  American Double / Imperial Stout            5.0           4.5   
591818  American Double / Imperial Stout            4.5           5.0   
591819  American Double / Imperial Stout            4.5           5.0   
591824  American Double / Imperial Stout            5.0           5.0   
...                                  ...            ...           ...   
653706          American Amber / Red Ale            2.5           3.5   
37234                     Czech Pilsener            3.0           2.0   
653695                        Hefeweizen            2.5           2.0   
39667            American Pale Ale (APA)            2.0           2.0   
39648                    Irish Dry Stout            2.0           2.0   

                         beer_name  beer_abv  beer_beerid  
793307          Fire Rock Pale Ale       5.8         5753  
591801  Bourbon County Brand Stout      14.5        10672  
591818  Bourbon County Brand Stout      14.5        10672  
591819  Bourbon County Brand Stout      14.5        10672  
591824  Bourbon County Brand Stout      14.5        10672  
...                            ...       ...          ...  
653706          Red Rock Amber Ale       4.0         3787  
37234                     Pub Pils       4.0         3806  
653695     Red Rock Bavarian Weiss       4.0         3788  
39667                Utah Pale Ale       4.0         3810  
39648            Latter Days Stout       4.0         3804  

[1586614 rows x 13 columns]

In [ ]:
# Number of reviews for each Beer
print(df['beer_name'].value_counts())

90 Minute IPA                          3290
India Pale Ale                         3130
Old Rasputin Russian Imperial Stout    3111
Sierra Nevada Celebration Ale          3000
Two Hearted Ale                        2728
                                       ... 
Titanbraü Pils                            1
Titanbraü Weizen                          1
Titanbräu Ale                             1
Texas Brunette                            1
Sausa Weizen                              1
Name: beer_name, Length: 56857, dtype: int64


In [ ]:
# Adding a column for No. of Reviews
df['no_of_reviews'] = df.groupby('beer_name')['beer_name'].transform('size')


         brewery_id                 brewery_name  review_time  review_overall  \
0             10325              Vecchio Birraio   1234817823             1.5   
1             10325              Vecchio Birraio   1235915097             3.0   
2             10325              Vecchio Birraio   1235916604             3.0   
3             10325              Vecchio Birraio   1234725145             3.0   
4              1075      Caldera Brewing Company   1293735206             4.0   
...             ...                          ...          ...             ...   
1586609       14359  The Defiant Brewing Company   1162684892             5.0   
1586610       14359  The Defiant Brewing Company   1161048566             4.0   
1586611       14359  The Defiant Brewing Company   1160702513             4.5   
1586612       14359  The Defiant Brewing Company   1160023044             4.0   
1586613       14359  The Defiant Brewing Company   1160005319             5.0   

         review_aroma  revi

#Calculating average of overall review value for each beer

In [ ]:
#adding a column for average of overall reviews for each beer
 df['average_overall_review']= df.groupby('beer_name')['review_overall'].transform('mean')

In [ ]:
df.head(10)

brewery_id             brewery_name  review_time  review_overall  \
0       10325          Vecchio Birraio   1234817823             1.5   
1       10325          Vecchio Birraio   1235915097             3.0   
2       10325          Vecchio Birraio   1235916604             3.0   
3       10325          Vecchio Birraio   1234725145             3.0   
4        1075  Caldera Brewing Company   1293735206             4.0   
5        1075  Caldera Brewing Company   1325524659             3.0   
6        1075  Caldera Brewing Company   1318991115             3.5   
7        1075  Caldera Brewing Company   1306276018             3.0   
8        1075  Caldera Brewing Company   1290454503             4.0   
9        1075  Caldera Brewing Company   1285632924             4.5   

   review_aroma  review_appearance review_profilename  \
0           2.0                2.5            stcules   
1           2.5                3.0            stcules   
2           2.5                3.0            stcules   
3           3.0                3.5            stcules   
4           4.5                4.0     johnmichaelsen   
5           3.5                3.5            oline73   
6           3.5                3.5          Reidrover   
7           2.5                3.5       alpinebryant   
8           3.0                3.5      LordAdmNelson   
9           3.5                5.0       augustgarage   

                       beer_style  review_palate  review_taste  \
0                      Hefeweizen            1.5           1.5   
1              English Strong Ale            3.0           3.0   
2          Foreign / Export Stout            3.0           3.0   
3                 German Pilsener            2.5           3.0   
4  American Double / Imperial IPA            4.0           4.5   
5            Herbed / Spiced Beer            3.0           3.5   
6            Herbed / Spiced Beer            4.0           4.0   
7            Herbed / Spiced Beer            2.0           3.5   
8            Herbed / Spiced Beer            3.5           4.0   
9            Herbed / Spiced Beer            4.0           4.0   

                beer_name  beer_abv  beer_beerid  no_of_reviews  \
0            Sausa Weizen       5.0        47986              1   
1                Red Moon       6.2        48213              1   
2  Black Horse Black Beer       6.5        48215              1   
3              Sausa Pils       5.0        47969              1   
4           Cauldron DIPA       7.7        64883              1   
5     Caldera Ginger Beer       4.7        52159              9   
6     Caldera Ginger Beer       4.7        52159              9   
7     Caldera Ginger Beer       4.7        52159              9   
8     Caldera Ginger Beer       4.7        52159              9   
9     Caldera Ginger Beer       4.7        52159              9   

   average_overall_review  
0                1.500000  
1                3.000000  
2                3.000000  
3                3.000000  
4                4.000000  
5                3.833333  
6                3.833333  
7                3.833333  
8                3.833333  
9                3.833333

#Sorting the List by average_overall_review for Beers with over 1000 reviews

In [ ]:
df[df['no_of_reviews'] > 1000].sort_values(by=['average_overall_review'], ascending=False).drop_duplicates('beer_name').head(20)


brewery_id                                       brewery_name  \
1490812         313  Brouwerij Westvleteren (Sint-Sixtusabdij van W...   
1134948         863                      Russian River Brewing Company   
691872          252            Bayerische Staatsbrauerei Weihenstephan   
1439278          63                                  Deschutes Brewery   
615896          199                      Ballast Point Brewing Company   
1141676         863                      Russian River Brewing Company   
33542          1199                           Founders Brewing Company   
1332544          26                 Three Floyds Brewing Co. & Brewpub   
410932          694                             Tröegs Brewing Company   
931216          287                               Bell's Brewery, Inc.   
1138139         863                      Russian River Brewing Company   
23003          1199                           Founders Brewing Company   
970576          222                        Brouwerij Duvel Moortgat NV   
340785          207                             Brasserie de Rochefort   
782325        13014                              Surly Brewing Company   
1339922          26                 Three Floyds Brewing Co. & Brewpub   
422126           73                        Great Lakes Brewing Company   
934315          287                               Bell's Brewery, Inc.   
1334865          26                 Three Floyds Brewing Co. & Brewpub   
1500296         259                         Brouwerij St. Bernardus NV   

         review_time  review_overall  review_aroma  review_appearance  \
1490812   1083719379             4.5           4.5                5.0   
1134948   1119155752             4.5           5.0                4.0   
691872    1162004611             5.0           4.5                4.0   
1439278   1189310459             4.5           4.5                5.0   
615896    1270882722             4.0           5.0                4.5   
1141676   1275598823             4.0           4.5                4.0   
33542     1238299872             4.5           4.0                4.5   
1332544   1150130812             4.5           4.0                4.0   
410932    1296600311             5.0           4.5                4.0   
931216    1246647242             4.5           5.0                4.5   
1138139   1288147375             4.0           4.5                4.5   
23003     1109640854             3.5           4.5                4.5   
970576    1067738926             4.5           4.5                5.0   
340785    1186283068             4.5           4.0                5.0   
782325    1302053329             4.5           4.0                3.5   
1339922   1259602822             4.0           4.5                4.0   
422126    1309397827             4.5           4.0                4.5   
934315    1200878460             4.5           4.0                4.5   
1334865   1322412599             4.5           4.5                4.0   
1500296   1082150536             4.5           4.5                4.5   

        review_profilename                        beer_style  review_palate  \
1490812             jvajda                  Quadrupel (Quad)            4.5   
1134948                ADR    American Double / Imperial IPA            4.0   
691872     mrasskicktastic                        Hefeweizen            4.5   
1439278    objectivemonkey  American Double / Imperial Stout            4.5   
615896                Pahn                      American IPA            4.0   
1141676       homebrewhawk                 American Wild Ale            4.5   
33542         dananabanana  American Double / Imperial Stout            4.5   
1332544              Brent           American Pale Wheat Ale            4.5   
410932            larryi86          American Amber / Red Ale            4.0   
931216          bonsreeb85    American Double / Imperial IPA            4.0   
1138139             dbhall                      Ame

# Top 3 Beers with over 1k reviews:
# 1. Trappist Westvleteren 12: 4.6179
# 2. Pliny The Elder: 4.5900
# 3. Weihenstephaner Hefeweissbier: 4.5159

# **I enjoy a beer which aroma and appearance matches the beer style. What beer should I buy?**

#Calculating average of reviews on aroma

In [ ]:
df['average_aroma_review'] = df.groupby('beer_name')['review_aroma'].transform('mean')

In [ ]:
df.head(5)

brewery_id             brewery_name  review_time  review_overall  \
0       10325          Vecchio Birraio   1234817823             1.5   
1       10325          Vecchio Birraio   1235915097             3.0   
2       10325          Vecchio Birraio   1235916604             3.0   
3       10325          Vecchio Birraio   1234725145             3.0   
4        1075  Caldera Brewing Company   1293735206             4.0   

   review_aroma  review_appearance review_profilename  \
0           2.0                2.5            stcules   
1           2.5                3.0            stcules   
2           2.5                3.0            stcules   
3           3.0                3.5            stcules   
4           4.5                4.0     johnmichaelsen   

                       beer_style  review_palate  review_taste  \
0                      Hefeweizen            1.5           1.5   
1              English Strong Ale            3.0           3.0   
2          Foreign / Export Stout            3.0           3.0   
3                 German Pilsener            2.5           3.0   
4  American Double / Imperial IPA            4.0           4.5   

                beer_name  beer_abv  beer_beerid  average_aroma_review  
0            Sausa Weizen       5.0        47986                   2.0  
1                Red Moon       6.2        48213                   2.5  
2  Black Horse Black Beer       6.5        48215                   2.5  
3              Sausa Pils       5.0        47969                   3.0  
4           Cauldron DIPA       7.7        64883                   4.5

In [ ]:
# Sorting in descending order by average aroma review
df.sort_values(by=['average_aroma_review'], ascending=False).drop_duplicates('beer_name').head(5)


brewery_id                             brewery_name  review_time  \
432970        22319                         Figure 8 Brewing   1309976783   
180937          396                 AleSmith Brewing Company   1308184585   
383156          387                   Edsten Brewing Company   1040430968   
188345          736                        Hurlimann Brewery   1101881831   
1065156        2092  Wiltse's Brew Pub and Family Restaurant   1323502499   

         review_overall  review_aroma  review_appearance review_profilename  \
432970              5.0           5.0                5.0       ModemMisuser   
180937              5.0           5.0                5.0     DarthKostrizer   
383156              5.0           5.0                5.0         beerguy101   
188345              4.0           5.0                4.5              lou91   
1065156             4.5           5.0                5.0      groovemonster   

                               beer_style  review_palate  review_taste  \
432970                American Barleywine            5.0           5.0   
180937   American Double / Imperial Stout            5.0           5.0   
383156            Belgian Strong Pale Ale            5.0           5.0   
188345                         Doppelbock            4.0           4.5   
1065156                    American Stout            5.0           5.0   

                                              beer_name  beer_abv  \
432970                       Date Night With Jumbo Love       NaN   
180937   Alesmith Speedway Stout - Vanilla And Coconut       12.0   
383156                                Edsten Triple-Wit      10.0   
188345                                 Samichlaus Brown      14.0   
1065156                          Wiltse's Blue Ox Stout       NaN   

         beer_beerid  average_aroma_review  average_appearance_review  
432970         70781                   5.0                      5.000  
180937         70054                   5.0                      4.625  
383156          1734                   5.0                      5.000  
188345          6527                   5.0                      4.750  
1065156        35488                   5.0                      5.000

#Calculating average of reviews on appearance

In [ ]:
df['average_appearance_review'] = df.groupby('beer_name')['review_appearance'].transform('mean')

In [ ]:
df.head(5)

brewery_id             brewery_name  review_time  review_overall  \
0       10325          Vecchio Birraio   1234817823             1.5   
1       10325          Vecchio Birraio   1235915097             3.0   
2       10325          Vecchio Birraio   1235916604             3.0   
3       10325          Vecchio Birraio   1234725145             3.0   
4        1075  Caldera Brewing Company   1293735206             4.0   

   review_aroma  review_appearance review_profilename  \
0           2.0                2.5            stcules   
1           2.5                3.0            stcules   
2           2.5                3.0            stcules   
3           3.0                3.5            stcules   
4           4.5                4.0     johnmichaelsen   

                       beer_style  review_palate  review_taste  \
0                      Hefeweizen            1.5           1.5   
1              English Strong Ale            3.0           3.0   
2          Foreign / Export Stout            3.0           3.0   
3                 German Pilsener            2.5           3.0   
4  American Double / Imperial IPA            4.0           4.5   

                beer_name  beer_abv  beer_beerid  average_aroma_review  \
0            Sausa Weizen       5.0        47986                   2.0   
1                Red Moon       6.2        48213                   2.5   
2  Black Horse Black Beer       6.5        48215                   2.5   
3              Sausa Pils       5.0        47969                   3.0   
4           Cauldron DIPA       7.7        64883                   4.5   

   average_appearance_review  
0                        2.5  
1                        3.0  
2                        3.0  
3                        3.5  
4                        4.0

In [ ]:
# Sorting in descending order by average appearance review
df.sort_values(by=['average_appearance_review'], ascending=False).drop_duplicates('beer_name').head(5)


brewery_id                brewery_name  review_time  review_overall  \
178768        24453         Intuition Ale Works   1316654727             4.5   
1202148       19146           Browar Konstancin   1232270653             2.5   
679198         9568         Teme Valley Brewery   1104784112             4.0   
477910         6811  Heidelberger Brauerei GmbH   1192985360             3.0   
1471185        1599    Prescott Brewing Company   1078700122             5.0   

         review_aroma  review_appearance review_profilename  \
178768            3.5                5.0        kylehay2004   
1202148           2.5                5.0               atis   
679198            4.0                5.0         lakevalley   
477910            3.5                5.0           TheSarge   
1471185           4.0                5.0       funkendunkel   

                     beer_style  review_palate  review_taste  \
178768       American Black Ale            4.0           4.5   
1202148         German Pilsener            3.0           3.0   
679198            Winter Warmer            4.5           4.5   
477910          German Pilsener            3.0           3.0   
1471185  Fruit / Vegetable Beer            4.0           4.5   

                         beer_name  beer_abv  beer_beerid  \
178768               TCB Black Rye       6.8        73328   
1202148  Zdrojowe Niepasteryzowane       4.4        47353   
679198                    Wass Ale       6.0        21399   
477910          Braustüb´l Pilsner       4.8        39262   
1471185                Bumbleberry       NaN         4448   

         average_aroma_review  average_appearance_review  
178768                    3.5                        5.0  
1202148                   2.5                        5.0  
679198                    4.0                        5.0  
477910                    3.5                        5.0  
1471185                   4.0                        5.0

#Aggregating Columns

In [ ]:
df.agg({'average_aroma_review': 'mean', 'average_appearance_review': 'mean'})

average_aroma_review         3.735636
average_appearance_review    3.841642
dtype: float64